### PR TITLE
internal: Work through temporarily allowed clippy lints, part 2

### PR DESCRIPTION
**Pull request by Urhengulas**
Another follow-up to https://github.com/rust-lang/rust-analyzer/pull/16401.

### CODE REVIEW

**Review by Veykril — APPROVED** [review]
Seems like an import is missing or something

**Comment by Urhengulas** [comment]
> Seems like an import is missing or something

Fixed

File: crates/hir-def/src/hir/format_args.rs [hunk 1]
```diff
@@ -177,9 +177,9 @@ pub(crate) fn parse(
     let text = s.text_without_quotes();
     let str_style = match s.quote_offsets() {
         Some(offsets) => {
-            let raw = u32::from(offsets.quotes.0.len()) - 1;
+            let raw = usize::from(offsets.quotes.0.len()) - 1;
             // subtract 1 for the `r` prefix
-            (raw != 0).then(|| raw as usize - 1)
```
Comment (Veykril) [inline crates/hir-def/src/hir/format_args.rs:182]: This is no longer lazy, so this always underflows when raw = 0

Comment (Veykril) [reply crates/hir-def/src/hir/format_args.rs:182]: Seems weird for clippy to lint this as arithmetic is side effectful in debug builds

File: crates/ide-completion/src/context/analysis.rs [hunk 2]
```diff
@@ -1267,8 +1267,7 @@ fn pattern_context_for(
     pat
         .syntax()
         .ancestors()
-        .skip_while(|it| ast::Pat::can_cast(it.kind()))
-        .next()
```
Comment (Veykril) [inline crates/ide-completion/src/context/analysis.rs:1271]: This is wrong

File: crates/hir-def/src/hir/format_args.rs (hunk 1 above)
Comment (Urhengulas) [reply crates/hir-def/src/hir/format_args.rs:182]: Yes, I am also a bit disappointed. But I guess in 99% of the cases doing it not lazy will have the correct behavior.

File: crates/flycheck/src/lib.rs [hunk 3]
```diff
@@ -538,7 +540,7 @@ impl CargoActor {
 }
 
 enum CargoMessage {
-    CompilerArtifact(cargo_metadata::Artifact),
+    CompilerArtifact(Box<cargo_metadata::Artifact>),
```
Comment (lnicola) [inline crates/flycheck/src/lib.rs:543]: This came up a couple of times before and I think bjorn3 pointed put that most of the time we're going to have this variant, so making this change will cause a lot of allocations.

I don't know if the cost of moving the large struct around is lower or higher than that of the allocations, though.

File: crates/ide-completion/src/context/analysis.rs (hunk 2 above)
Comment (Veykril) [reply crates/ide-completion/src/context/analysis.rs:1271]: Oh I need to get my eyes checked, I thought the negation was missing, nevermind then!

File: crates/flycheck/src/lib.rs (hunk 3 above)
Comment (Urhengulas) [reply crates/flycheck/src/lib.rs:543]: I am fine with any choice. I was just going with the clippy recommendation here.

I'd propose that we go with it and if someone takes the time to compare the two options we potentially revert it and disable the clippy lint for it.

Comment (Veykril) [reply crates/flycheck/src/lib.rs:543]: https://github.com/rust-lang/rust-analyzer/pull/16470

**Pull request merged**